include/pybind11/cast.h

```diff
@@ -78,20 +81,152 @@ PYBIND11_NOINLINE inline internals &get_internals() {
         );
         internals_ptr->static_property_type = make_static_property_type();
         internals_ptr->default_metaclass = make_default_metaclass();
+        internals_ptr->instance_base = make_object_base_type(internals_ptr->default_metaclass);
     }
     return *internals_ptr;
 }
 
+/** Iterator that starts from a python base class and iterators through all pybind-registered bases
+ * of that class.  If not using Python-side multiple inheritance, this iterator iterates at most
+ * once.
+ */
+class type_info_iterator {
```

Comment:
The latest push addresses this basically as you suggested, with a much simpler iterator implementation (i.e. the complex logic is moved to the container class rather than implemented in the iterator).